### PR TITLE
scripts: aarch64: Reduce source code build in CI

### PR DESCRIPTION
**Pull request by MrXinWang**
As Linux kernel, edk2, virtiofd, SPDK are all build from source, we
need to find a way to reduce the number of source code build.

To address this issue, this commit adds a stage that touching a
".built" file to each repo of the source code after a successfully
build. By checking if the ".built" file exists, we can determine
if there is already an old build. Since the source code repo will
be removed and cloned again when we bump the repo hash, the CI will
rebuild the repo.

Signed-off-by: Henry Wang <Henry.Wang@arm.com>

### CODE REVIEW

File: scripts/run_integration_tests_aarch64.sh [hunk 1]
```diff
@@ -252,11 +252,21 @@ update_workloads() {
         echo "bar" > "$SHARED_DIR/file3" || exit 1
     fi
 
-    # checkout and build SPDK NVMe
-    build_spdk_nvme
+    # Check and build SPDK NVMe
+    NVMF_TGT="$WORKLOADS_DIR/spdk-nvme/nvmf_tgt"
+    RPC_PY="$WORKLOADS_DIR/spdk-nvme/rpc.py"
+    RPC="$WORKLOADS_DIR/spdk-nvme/rpc"
+    if [[ ! -f "$NVMF_TGT" || ! -f "$RPC_PY" || ! -d "$RPC" ]]; then
```
Comment (michael2012z) [inline scripts/run_integration_tests_aarch64.sh:259]: This logic doesn't work correctly if the source code version is bumped.

File: scripts/run_integration_tests_aarch64.sh [hunk 2]
```diff
@@ -252,11 +252,21 @@ update_workloads() {
...
+        build_spdk_nvme
+    else
+        cp -r "$WORKLOADS_DIR/spdk-nvme" /usr/local/bin/spdk-nvme
+    fi
 
     # Check and build EDK2 binary
-    build_edk2
+    EDK2_CLOUDHV_EFI="$WORKLOADS_DIR/CLOUDHV_EFI.fd"
+    if [ ! -f "$EDK2_CLOUDHV_EFI" ]; then
```
Comment (michael2012z) [inline scripts/run_integration_tests_aarch64.sh:267]: Same.

**Review by michael2012z — CHANGES_REQUESTED** [review]

File: scripts/run_integration_tests_aarch64.sh (hunk 1 above)
Comment (MrXinWang) [reply scripts/run_integration_tests_aarch64.sh:259]: indeed, I will fix that

Comment (MrXinWang) [reply scripts/run_integration_tests_aarch64.sh:261]: done

File: scripts/run_integration_tests_aarch64.sh (hunk 2 above)
Comment (MrXinWang) [reply scripts/run_integration_tests_aarch64.sh]: done.

**Comment by jongwu** [comment]
I think it's better to checkout to a stable tag not a certain commit.

File: scripts/run_integration_tests_aarch64.sh [hunk 3]
```diff
@@ -92,33 +95,62 @@ build_edk2() {
...
+        pushd $SPDK_DIR
+        git submodule update --init
+        apt-get update
+        ./scripts/pkgdep.sh
+        ./configure --with-vfio-user
+        chmod +x /usr/local/lib/python3.8/dist-packages/ninja/data/bin/ninja
+        make -j `nproc` || exit 1
+        touch .built.spdk
+        popd
+    else
```
Comment (michael2012z) [inline scripts/run_integration_tests_aarch64.sh:131]: If it is the first build, the files below cannot be copied.
You should remove the `else` condition and move the instructions in it outside.

File: scripts/run_integration_tests_aarch64.sh [hunk 4]
```diff
@@ -59,13 +59,16 @@ build_custom_linux() {
...
-    cp arch/arm64/boot/Image.gz "$WORKLOADS_DIR/" || exit 1
-    popd
+    if [ ! -f "$LINUX_CUSTOM_DIR/.built.linux-custom" ]; then
+        cp $SRCDIR/resources/linux-config-aarch64 $LINUX_CUSTOM_DIR/.config
+
+        pushd $LINUX_CUSTOM_DIR
+        time make -j `nproc`
+        cp arch/arm64/boot/Image "$WORKLOADS_DIR/" || exit 1
+        cp arch/arm64/boot/Image.gz "$WORKLOADS_DIR/" || exit 1
+        touch .built.linux-custom
```
Comment (michael2012z) [inline scripts/run_integration_tests_aarch64.sh:69]: You don't need to give different names to such indicators. Using an uniformed name `.built` could be simply and clean.

**Review by michael2012z — CHANGES_REQUESTED** [review]

File: scripts/run_integration_tests_aarch64.sh (hunk 3 above)
Comment (MrXinWang) [reply scripts/run_integration_tests_aarch64.sh]: Done

File: scripts/run_integration_tests_aarch64.sh (hunk 4 above)
Comment (MrXinWang) [reply scripts/run_integration_tests_aarch64.sh]: Done

**Comment by MrXinWang** [comment]
> I think it's better to checkout to a stable tag not a certain commit.

Sure, I will do it in a separate PR.

**Review by michael2012z — APPROVED** [review]

File: scripts/run_integration_tests_aarch64.sh [hunk 5]
```diff
@@ -59,13 +59,16 @@ build_custom_linux() {
...
     checkout_repo "$LINUX_CUSTOM_DIR" "$LINUX_CUSTOM_URL" "$LINUX_CUSTOM_BRANCH"
 
-    cp $SRCDIR/resources/linux-config-aarch64 $LINUX_CUSTOM_DIR/.config
-
-    pushd $LINUX_CUSTOM_DIR
-    time make -j `nproc`
-    cp arch/arm64/boot/Image "$WORKLOADS_DIR/" || exit 1
-    cp arch/arm64/boot/Image.gz "$WORKLOADS_DIR/" || exit 1
-    popd
+    if [ ! -f "$LINUX_CUSTOM_DIR/.built" ]; then
```
Comment (rbradford) [inline scripts/run_integration_tests_aarch64.sh:62]: This seems like a bad idea. The kernel will not get rebuilt on a config change? Or depending on `checkout_repo` if the repository changes.

File: scripts/run_integration_tests_aarch64.sh [hunk 6]
```diff
@@ -217,19 +248,7 @@ update_workloads() {
...
-        git clone "https://gitlab.com/virtio-fs/virtiofsd.git" $VIRTIOFSD_DIR
-        pushd $VIRTIOFSD_DIR
-        git checkout v1.1.0
-        time cargo build --release
-        cp target/release/virtiofsd $VIRTIOFSD || exit 1
-        popd
-        rm -rf $VIRTIOFSD_DIR
-        popd
-    fi
+    build_virtiofsd
```
Comment (rbradford) [inline scripts/run_integration_tests_aarch64.sh:251]: I thought this was already removed. We are only testing Rust based virtiofsd.

Comment (MrXinWang) [reply scripts/run_integration_tests_aarch64.sh:251]: Yes, the C version virtiofsd was removed in the dockerfile. But I think the rust version is built in both x86 and arm integration test scripts

Comment (rbradford) [reply scripts/run_integration_tests_aarch64.sh:251]: Whoops. Apologires I forgot that the Rust version was renamed. You're spot on. Does this code correctly rebuild if the commit hash is changed?

Comment (MrXinWang) [reply scripts/run_integration_tests_aarch64.sh:251]: Yes, if the hash is changed, the checkout_repo will delete the source dir and therefore lead to a new build of the binary.

File: scripts/run_integration_tests_aarch64.sh (hunk 5 above)
Comment (michael2012z) [reply scripts/run_integration_tests_aarch64.sh:62]: Aha, yes this is a mistake.

In retrospect, I merged this PR too early to collect enough feedback.

**Pull request merged**